util: Set safe permissions for data directory and `wallets/` subdir

hebasto opened the pull request:
On master (1e7564eca8a688f39c75540877ec3bdfdde766b1) docs say:
```
$ ./src/bitcoind -help | grep -A 3 sysperms
  -sysperms
       Create new files with system default permissions, instead of umask 077
       (only effective with disabled wallet functionality)

```

Basing on that, one could expect that running `bitcoind` first time will create data directory and `wallets/` subdirectory with safe 0700 permissions.

But that is not the case:
```
$ stat .bitcoin | grep id
Access: (0775/drwxrwxr-x)  Uid: ( 1000/ hebasto)   Gid: ( 1000/ hebasto)
$ stat .bitcoin/wallets | grep id
Access: (0775/drwxrwxr-x)  Uid: ( 1000/ hebasto)   Gid: ( 1000/ hebasto)
```

Both directories, in fact, are created with system default permissions.

With this PR:
```
$ stat .bitcoin/wallets | grep id
Access: (0700/drwx------)  Uid: ( 1000/ hebasto)   Gid: ( 1000/ hebasto)
$ stat .bitcoin/wallets | grep id
Access: (0700/drwx------)  Uid: ( 1000/ hebasto)   Gid: ( 1000/ hebasto)
```

---

This PR:
- is alternative to bitcoin/bitcoin#13389 
- fixes bitcoin/bitcoin#15902 
- fixes bitcoin/bitcoin#22595
- closes bitcoin/bitcoin#13371
- reverts bitcoin/bitcoin#4286

Changes in behavior: removed `-sysperms` command-line argument / configure option. The related discussions are here:
- https://github.com/bitcoin/bitcoin/pull/13389#issuecomment-395306690
- https://github.com/bitcoin/bitcoin/pull/13389#issuecomment-539906114
- https://github.com/bitcoin/bitcoin/pull/13389#discussion_r279160472

If users rely on non-default access permissions, they could use `chmod`.

laanwj commented on src/util/system.cpp:
```diff
@@ -741,6 +750,10 @@ const fs::path &GetBlocksDir()
 
 const fs::path &GetDataDir(bool fNetSpecific)
 {
+#ifndef WIN32
```
GetDataDir, a getter, is not the place to do things such as set the umask. Please don't add these kind of side-effects.

hebasto replied:
Maybe `GetDataDir` is not a proper name for this function as it already has an intended side-effect:
https://github.com/bitcoin/bitcoin/blob/ddd3cd30cb050fe9b8738778beecba1d8bd50d52/src/util/system.cpp#L777-L780

laanwj replied:
I know. I kind of want to get rid of that too. But don't add any more.

hebasto replied:
Done.

laanwj commented:
Concept ACK on removing `-sysperms`.

hebasto commented:
@laanwj 
Thank you for your review. All your comments have been addressed.

hebasto commented on src/util/system.cpp:
```diff
@@ -65,6 +65,15 @@
 #include <thread>
 
+#ifndef WIN32
+static const int private_umask = 077;
```
Done.

laanwj commented on src/util/system.h:
```diff
@@ -35,6 +35,10 @@
 #include <boost/thread/condition_variable.hpp> // for boost::thread_interrupted
 
+#ifndef WIN32
+void SetDefaultUmask();
```
Wouldn't `SetupEnvironment` be the right place for this? Or is that too early?

hebasto replied:
Currently, `SetDefaultUmask()` is called just before the first possible disk write.

`SetupEnvironment()` seems too broad, as it is called in utilities too. And, yes, I think it is too early. But I could be wrong ;)

laanwj replied:
But does that matter? What would be the consequences of setting it too early? Why wouldn't we want to set the umask for the utilties?

hebasto replied:
> Wouldn't `SetupEnvironment` be the right place for this?

Agree.

hebasto commented:
@laanwj Thank you for review. Your comment has been addressed.

hebasto commented:
@promag Would you mind reviewing this PR?

hebasto commented:
Rebased 7213e8cbe685556365021176d686f45bdd036dae -> 67b12091c78b0183aa9dc451d176b16f832d93f7 ([pr17127.04](https://github.com/hebasto/bitcoin/commits/pr17127.04) -> [pr17127.05](https://github.com/hebasto/bitcoin/commits/pr17127.05)) due to the conflict with #15761.

hebasto commented:
Rebased 67b12091c78b0183aa9dc451d176b16f832d93f7 -> 297e61cdb8293c0bfd1d43f22eefbece3f541802 ([pr17127.05](https://github.com/hebasto/bitcoin/commits/pr17127.05) -> [pr17127.06](https://github.com/hebasto/bitcoin/commits/pr17127.06)) due to the conflict with #16224.

jonasschnelli commented:
utACK 297e61cdb8293c0bfd1d43f22eefbece3f541802

hebasto commented:
Rebased 297e61cdb8293c0bfd1d43f22eefbece3f541802 -> 337a7709502f562b764cce78f6de1d1a744cebd8 ([pr17127.06](https://github.com/hebasto/bitcoin/commits/pr17127.06) -> [pr17127.07](https://github.com/hebasto/bitcoin/commits/pr17127.07)) due to the conflict with #15935.

hebasto commented:
Rebased 337a7709502f562b764cce78f6de1d1a744cebd8 -> 788143749f7a71d0b3af12275e2fde7dc74ff03d ([pr17127.07](https://github.com/hebasto/bitcoin/commits/pr17127.07) -> [pr17127.08](https://github.com/hebasto/bitcoin/commits/pr17127.08)) due to the conflict with #19561.

hebasto commented:
Rebased 788143749f7a71d0b3af12275e2fde7dc74ff03d -> 677617651e5b831da8d12f769a47d2957400d987 ([pr17127.08](https://github.com/hebasto/bitcoin/commits/pr17127.08) -> [pr17127.09](https://github.com/hebasto/bitcoin/commits/pr17127.09)) due to the conflict with #19779.

hebasto commented:
Rebased 677617651e5b831da8d12f769a47d2957400d987 -> ef8ed07715f90bc156a4d4015ea1db9ee975963d ([pr17127.09](https://github.com/hebasto/bitcoin/commits/pr17127.09) -> [pr17127.10](https://github.com/hebasto/bitcoin/commits/pr17127.10)) due to the conflict with #15367.

practicalswift commented:
Concept ACK

hebasto commented:
Rebased ef8ed07715f90bc156a4d4015ea1db9ee975963d -> 9b105d7289a793eaa30ae8fe7937b1356969758b ([pr17127.10](https://github.com/hebasto/bitcoin/commits/pr17127.10) -> [pr17127.11](https://github.com/hebasto/bitcoin/commits/pr17127.11)) due to the conflict with #20744.

hebasto commented:
PR description has been updated.

willcl-ark reviewed (approved):
ACK 9b105d7289a793eaa30ae8fe7937b1356969758b

The test datadir will have the following files created in it after node shutdown:

banlist.json, debug.log, fee_estimates.dat, mempool.dat, peers.dat, settings.json

Would it be worth adding a permissions test for e.g debug.log to check that the correct umask is also being applied to files?

willcl-ark commented:
> Sorry, but I didn't quite get your suggestion. Mind elaborating it?

My fault for not being clearer!

You added a test for [directory permissions](https://github.com/hebasto/bitcoin/blob/20191013-permissions/test/functional/feature_fs_permissions.py#L19-L23), but not not one for files. As files and folders will be created with different permissions,  it might make sense to add this at the same time?

For directories the test covers `datadir` and `walletsdir`. For files I was suggesting we could test `debug.log`.

willcl-ark commented:
This also fixes https://github.com/bitcoin/bitcoin/issues/22595

hebasto commented:
Updated 9b105d7289a793eaa30ae8fe7937b1356969758b -> 07c496d01b7ab03451b1d8278cef2652540ff191 ([pr17127.11](https://github.com/hebasto/bitcoin/commits/pr17127.11) -> [pr17127.12](https://github.com/hebasto/bitcoin/commits/pr17127.12)):
- rebased
- addressed @willcl-ark's [comment](https://github.com/bitcoin/bitcoin/pull/17127#issuecomment-1332405771)

willcl-ark reviewed (approved):
ACK 07c496d01

Might want to add this to the release notes too?

fanquake commented on src/util/system.cpp:
```diff
@@ -64,6 +64,16 @@
 
+#ifndef WIN32
+namespace {
+constexpr mode_t private_umask = 0077;
```
Is there a reason to add 10 lines here, and another `#ifndef`, instead of putting the 2 lines of code where they are actually used? i.e why not just:
```cpp
#ifndef WIN32
    constexpr mode_t private_umask = 0077;
    umask(private_umask);
#endif
```
in `SetupEnvironment()`?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/17127#issuecomment-1417887067).

hebasto committed:
Remove `-sysperms` option

This change effectively reverts commits from
https://github.com/bitcoin/bitcoin/pull/4286.

Users, who rely on non-default access permissions, should use `chmod`
command.

hebasto commented:
Updated 07c496d01b7ab03451b1d8278cef2652540ff191 -> 15c7105a2ddc720ed0cac0fe7995755cad437880 ([pr17127.12](https://github.com/hebasto/bitcoin/commits/pr17127.12) -> [pr17127.13](https://github.com/hebasto/bitcoin/commits/pr17127.13)):
- rebased
- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/17127#discussion_r1063355778)

john-moffett reviewed (commented):
Approach ACK 15c7105a2ddc720ed0cac0fe7995755cad437880

I agree that this may warrant an explicit release note considering a command line option is being dropped.

hebasto committed:
Apply default umask in `SetupEnvironment()`

This change makes all filesystem artifacts--files and directories--being
created with the default umask.

hebasto committed:
test: Add test for file system permissions

hebasto commented:
Updated 15c7105a2ddc720ed0cac0fe7995755cad437880 -> c9ba4f9ecb1a282d98e7456a84ca84362b161757 ([pr17127.13](https://github.com/hebasto/bitcoin/commits/pr17127.13) -> [pr17127.14](https://github.com/hebasto/bitcoin/commits/pr17127.14), [diff](https://github.com/hebasto/bitcoin/compare/pr17127.13..pr17127.14)):

- addressed @john-moffett's comments

hebasto commented on src/i2p.cpp:
```diff
@@ -336,7 +336,7 @@ void Session::GenerateAndSavePrivateKey(const Sock& sock)
     DestGenerate(sock);
 
-    // umask is set to 077 in init.cpp, which is ok (unless -sysperms is given)
+    // umask is set to 077 in init.cpp, which is ok.
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/17127#issuecomment-1418911048).

hebasto commented on src/rpc/request.cpp:
```diff
@@ -86,7 +86,7 @@ bool GenerateAuthCookie(std::string *cookie_out)
 
     /** the umask determines what permissions are used to create this file -
-     * these are set to 077 in init.cpp unless overridden with -sysperms.
+     * these are set to 077 in init.cpp.
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/17127#issuecomment-1418911048).

john-moffett commented:
ACK c9ba4f9ecb1a282d98e7456a84ca84362b161757

willcl-ark reviewed (approved):
ACK c9ba4f9ecb1a282d98e7456a84ca84362b161757
